openapi.yaml

```diff
@@ -4711,16 +4711,35 @@ components:
       example: null
       description: |-
         The namespace the `process_id` is valid for.
+
+        The following options are pre-defined by the openEO API, but additional
+        namespaces may be introduced by back-ends or in a future version of the API.
         
-        By default, the namespace is `null`, which includes all pre-defined processes listed at `GET /processes`.
-        The namespace `user` refers to all user-defined processes listed as `GET /process_graphs`.
-        Additional namespaces may be introduced by back-ends or in a future version of the API.
+        * `null` (default): Checks both user-defined and pre-defined processes,
+           but prefers user-defined processes if both are available.
+           This allows users to replace missing pre-defined processes with
+           user-defined processes for portability, e.g. common processes from
+           [processes.openeo.org](https://processes.openeo.org) that have a process graph
+           included.
+        * `backend`: Uses exclusively the pre-defined processes listed at `GET /processes`.
+        * `user`: Uses exclusively the user-defined processes listed at `GET /process_graphs`.
+
+        Note on collisions: The only way a collision between user-defined processes and
+        pre-defined processes can occur is when back-ends introduce a new process with an
+        identifier already used by users. User-defined processes with an identifier used by
+        a pre-defined process MUST be rejected. To ensure reproducibility, back-ends SHOULD
```

Comment:
I don't like this rule "collisions are ok if UDP was first". I don't think this will scale well if other namespaces are added. I would just don't care about "collisions" on this level because there is no collision: there are in different namespaces.

as noted by @m-mohr client could still raise a warning if they detect this kind of name collision when creating a UDP, which sound better because the API itesel has no "warning" channel

Reply:
> I don't like this rule "collisions are ok if UDP was first". I don't think this will scale well if other namespaces are added. I would just don't care about "collisions" on this level because there is no collision: there are in different namespaces.

This is meant to only apply between PDP and UDP. Other namespace are not involved.

> as noted by @m-mohr : the client could still raise a warning if they detect this kind of name collision when creating a UDP, which sound better because the API itself has no "warning" channel

It has the log files endpoints though.

Reply:
> It has the log files endpoints though

also for UDP creation?

Reply:
No, just for jobs/services indeed.

Reply:
I like the idea of having the client raising a warning, since users interface with that rather than the back-end

Reply:
> This is meant to only apply between PDP and UDP. Other namespace are not involved.

indeed, and that's why it does seem to be so arbitrary to me. Or you forbid all possible collisions, or you allow them everywhere, something in between feels hackish and ad-hoc.

Reply:
Okay, I'm going to replace the UDP rejection rule with the following:

> Clients SHOULD warn the user if a user-defined process is added with an identifier that exists in the pre-defined processes.

Reply:
See https://github.com/Open-EO/openeo-api/pull/313/commits/59703ae6d82faebd44b8951a735fac64bdf1a4b8